model/scratch/dffml_model_scratch/model/slr.py

```diff
@@ -0,0 +1,153 @@
+# SPDX-License-Identifier: MIT
+# Copyright (c) 2019 Intel Corporation
+'''
+Description of what this model does
+'''
+import os
+import abc
+import hashlib
+import json
+from typing import AsyncIterator, Tuple, Any, List, Optional, NamedTuple, Dict
+from statistics import mean
+import numpy as np
+
+from dffml.repo import Repo
+from dffml.source.source import Sources
+from dffml.feature import Features
+from dffml.accuracy import Accuracy
+from dffml.model.model import ModelConfig, ModelContext, Model
+from dffml.util.entrypoint import entry_point
+from dffml.util.cli.arg import Arg
+
+
+class SLRConfig(ModelConfig, NamedTuple):
+    predict: str
+    directory: str
+
+
+class SLRContext(ModelContext):
+    '''
+    Model wraping scratch API
+    '''
+    def __init__(self, parent):
+        super().__init__(parent)
+        self.xData = np.array([])
+        self.yData = np.array([])
+        self.key = None
+        self.regression_line = self.get_regression_line()
+
+    def get_regression_line(self):
+        return self.parent.saved.get(self.key, None)
+
+    async def set_regression_line(self, slope, constant, accuracy):
+        self.parent.saved[self.key] = (slope, constant, accuracy)
+
+    async def squared_error(self, ys, yline):
+        return sum((ys - yline)**2)
+
+    async def coeff_of_deter(self, ys, regression_line):
+        y_mean_line = [mean(ys) for y in ys]
+        squared_error_mean = await self.squared_error(ys, y_mean_line)
+        squared_error_regression = await self.squared_error(ys, regression_line)
+        return 1 - (squared_error_regression/squared_error_mean)
+
+    async def applicable_features(self, features: Features):
+        usable = []
+        if len(features) != 1:
+            raise ValueError("Simple Linear Regression doesn't support features other than 1")
+        for feature in features:
+            if feature.dtype() != int and feature.dtype() != float:
+                raise ValueError("Simple Linear Regression only supports int or float feature")
+            if feature.length() != 1:
+                raise ValueError("Simple LR only supports single values (non-matrix / array)")
+            usable.append(feature.NAME)
+        return usable
+
+    async def best_fit_line(self, xs, ys):
+        m = ((mean(xs)*mean(ys)) - mean(xs*ys))/((mean(xs)*mean(xs)) - (mean(xs*xs)))
```

Comment:
perf nit: assign temp vars to function calls and limit function call , matters when you scale and improves readability
```python
mean_xs = np.mean(xs)
mean_ys = np.mean(ys)
m = (mean_xs * mean_ys - np.mean(xs * ys))/(mean_xs ** 2 - np.mean(xs * xs))
b = mean_ys  - m * mean_xs
```

Reply:
you could even use [np.polyfit](https://docs.scipy.org/doc/numpy/reference/generated/numpy.polyfit.html) to find the intercept and bias, but if @pdxjohnny  wants the implementation to be vanilla without using things like `polyfit`, let's make sure everything is vectorized